orchestrator/careplanservice/handle_getcareplan.go

```diff
@@ -28,29 +33,112 @@ func (s *Service) handleGetCarePlan(ctx context.Context, id string, headers *fhi
-	if err != nil {
-		return nil, err
-	}
+	tx.Get(carePlan, "CarePlan/"+request.ResourceId, coolfhir.WithAuditEvent(ctx, tx, coolfhir.AuditEventInfo{
+		Action: fhir.AuditEventActionR,
+		ActingAgent: &fhir.Reference{
+			Identifier: &request.Principal.Organization.Identifier[0],
+			Type:       to.Ptr("Organization"),
+		},
+		Observer: *request.LocalIdentity,
+	}))
 
-	return &carePlan, nil
+	carePlanEntryIdx := 0
+
+	return func(txResult *fhir.Bundle) ([]*fhir.BundleEntry, []any, error) {
+		var retCarePlan fhir.CarePlan
+		result, err := coolfhir.NormalizeTransactionBundleResponseEntry(ctx, s.fhirClient, s.fhirURL, &tx.Entry[carePlanEntryIdx], &txResult.Entry[carePlanEntryIdx], &retCarePlan)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to process CarePlan read result: %w", err)
+		}
+		// We do not want to notify subscribers for a get
+		return []*fhir.BundleEntry{result}, []any{}, nil
+	}, nil
 }
 
 // handleSearchCarePlan does a search for CarePlan based on the user requester parameters. If CareTeam is not requested, add this to the fetch to be used for validation
 // if the requester is a participant of one of the returned CareTeams, return the whole bundle, else error
 // Pass in a pointer to a fhirclient.Headers object to get the headers from the fhir client request
-func (s *Service) handleSearchCarePlan(ctx context.Context, queryParams url.Values, headers *fhirclient.Headers) (*fhir.Bundle, error) {
-	// Verify requester is authenticated
-	principal, err := auth.PrincipalFromContext(ctx)
+func (s *Service) handleSearchCarePlan(ctx context.Context, request FHIRHandlerRequest, tx *coolfhir.BundleBuilder) (FHIRHandlerResult, error) {
+	log.Ctx(ctx).Info().Msgf("Searching for CarePlans")
+
+	bundle, err := s.searchCarePlan(ctx, request.QueryParams, request.FhirHeaders, *request.Principal)
 	if err != nil {
 		return nil, err
 	}
 
+	carePlanEntryIndexes := []int{}
+
+	for _, entry := range bundle.Entry {
+		var currentCarePlan fhir.CarePlan
+		if err := json.Unmarshal(entry.Resource, &currentCarePlan); err != nil {
+			log.Ctx(ctx).Error().
+				Err(err).
+				Msg("Failed to unmarshal resource for audit")
+			continue
+		}
+
+		// Create the query detail entity
+		queryEntity := fhir.AuditEventEntity{
+			Type: &fhir.Coding{
+				System:  to.Ptr("http://terminology.hl7.org/CodeSystem/audit-entity-type"),
+				Code:    to.Ptr("2"), // query parameters
+				Display: to.Ptr("Query Parameters"),
+			},
+			Detail: []fhir.AuditEventEntityDetail{},
+		}
+
+		// Add each query parameter as a detail
+		for param, values := range request.QueryParams {
+			queryEntity.Detail = append(queryEntity.Detail, fhir.AuditEventEntityDetail{
+				Type:        param, // parameter name as string
+				ValueString: to.Ptr(strings.Join(values, ",")),
+			})
+		}
+
+		carePlanEntryIndexes = append(carePlanEntryIndexes, len(tx.Entry))
+
+		tx.Get(entry.Resource, "CarePlan/"+*currentCarePlan.Id, coolfhir.WithAuditEvent(ctx, tx, coolfhir.AuditEventInfo{
```

Comment:
why read the CarePlan again if we already have it? Can't we just write the audit events?